File: src/eval/summary.cpp
#include "summary.hpp"
#include <algorithm>

namespace trafficsim {
void SummaryAccumulator::add(const SimEvent& event) {
    if (std::holds_alternative<SafetyClampEvent>(event)) ++clamps_;
    if (const auto* arrived = std::get_if<ArrivedEvent>(&event)) {
        ++completed_;
        travel_ += arrived->travelTime;
        delay_ += std::max(0.0, arrived->travelTime + arrived->departureDelay - arrived->freeFlowTime);
    }
}
RunSummary SummaryAccumulator::summary() const {
    RunSummary result{completed_, clamps_, {}, {}};
    if (completed_) {
        result.meanTravelTime = travel_ / static_cast<double>(completed_);
        result.meanDelay = delay_ / static_cast<double>(completed_);
    }
    return result;
}
std::size_t pendingCount(const SimState& state) {
    std::size_t count = 0;
    for (const auto& input : state.inputs) count += input.queue.size();
    return count;
}
}

File: src/eval/summary.ts
import type { SimEvent } from '../core';

export interface RunSummary {
  readonly validation: 'not-yet-validated';
  readonly completed: number;
  readonly safetyClamps: number;
  readonly meanTravelTime: number | null;
  /** Completed trips only; excess over constant desired-speed travel, including source wait. */
  readonly meanDelay: number | null;
}

/** M0 diagnostic, not HCM control delay or a LOS evaluation. */
export function summarize(events: Iterable<SimEvent>): RunSummary {
  let completed = 0;
  let safetyClamps = 0;
  let totalTravelTime = 0;
  let totalDelay = 0;
  for (const event of events) {
    if (event.kind === 'safety-clamp') safetyClamps++;
    if (event.kind !== 'arrived') continue;
    completed++;
    totalTravelTime += event.travelTime;
    totalDelay += Math.max(0, event.travelTime + event.departureDelay - event.freeFlowTime);
  }
  return {
    validation: 'not-yet-validated', completed, safetyClamps,
    meanTravelTime: completed ? totalTravelTime / completed : null,
    meanDelay: completed ? totalDelay / completed : null,
  };
}
